[FLINK-16641][network] Announce sender's backlog to solve the deadlock issue without exclusive buffers

wsry opened the pull request:
## What is the purpose of the change

This is the second ingredient besides FLINK-16404 to solve the deadlock problem without exclusive buffers.

The scenario is as follows:

The data in subpartition with positive backlog can be sent without doubt because the exclusive credits would be feedback finally.
Without exclusive buffers, the receiver would not request floating buffers for 0 backlog. But when the new backlog is added into such subpartition, it has no way to notify the receiver side without positive credits ATM.
So it would result in waiting for each other between receiver and sender sides to cause deadlock. The sender waits for credit to notify backlog and the receiver waits for backlog to request floating credits.
To solve the above problem, the sender needs a separate message to announce backlog sometimes besides existing `BufferResponse`. Then the receiver can get this info to request floating buffers to feedback.

In the current implementation, the upstream task will announce backlog to the downstream task and the downstream task will allocate buffers (credit) to receive data from upstream. This PR enhances the current implementation in three main aspects:
1. If there is no initial credit, the upstream producer task will announce the available backlog to the downstream consumer task when available data is notified.
2. The downstream consumer task will release all allocated buffers (credit) on receiving the aligned checkpoint barrier. Besides, it will never allocate any credit before checkpoint completion.
3. For empty buffers of upstream task, instead of released directly, they will be sent to the downstream task to release the buffers (credit) allocated for these empty buffers.


## Brief change log

  - If there is no initial credit, the upstream producer task will announce the available backlog to the downstream consumer task when available data is notified.
  - The downstream consumer task will release all allocated buffers (credit) on receiving the aligned checkpoint barrier. Besides, it will never allocate any credit before checkpoint completion.
  - For empty buffers of upstream task, instead of released directly, they will be sent to the downstream task to release the buffers (credit) allocated for these empty buffers.

## Verifying this change

This change is verified by both existing tests and newly added tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -428,6 +423,23 @@ public void notifyBufferDestroyed() {
 	}
 
+	@Override
+	public void onCheckpointBarrier(CheckpointBarrier barrier) {
```
Make the following condition out of `synchronized` part, then we do not need to touch the lock for most of the cases.
```
if (!(initialCredit == 0 && options.isExactlyOnceMode() && !options.isUnalignedCheckpoint()))  {
     return;
}
```

zhijiangW replied:
It might be better to trigger this action by netty thread when received this barrier immediately. Otherwise the task processing might delay much time to better reuse the floating buffers for other channels.

zhijiangW replied:
Considering the race condition between this operation and canceler task, it might involve in potential conflicts and deadlock case if `removeBufferListener` and adjust `isWaitingForFloatingBuffers` here. Let me think whether we can lazy remove the listener and adjust `isWaitingForFloatingBuffers` by reusing the existing process `RemoteInputChannel#notifyBufferAvailable`.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -234,11 +236,16 @@ private void writeAndFlushNextMessageIfPossible(final Channel channel) throws IO
-						reader.getReceiverId(),
-						next.buffersInBacklog());
+					Object msg;
+					if (next.buffer() != null) {
```
The different message path is distinguished in both `PartitionRequestQueue` and `CreditBasedSequenceNumberingViewReader` now. We can improve it to judge only in one place instead.

1. Introduce `ServerOutboundMessage` class to extend `NettyMessage` and make `AddBacklog` and `BufferResponse` both extend `ServerOutboundMessage`.
2. Introduce `NetworkSequenceViewReader#getNextMessage` instead of existing `NetworkSequenceViewReader#getNextBuffer`. And inside `CreditBasedSequenceNumberingViewReader` implementation, we can judge the condition for distinguish.

```
public NettyMessage.ServerOutboundMessage getNextMessage() throws IOException {
		if (numCreditsAvailable == 0 && initialCredit == 0 && !subpartitionView.isAvailable(numCreditsAvailable)) {
			return getBacklogMessage();
		} else {
			return getNextBufferResponse();
		}
	}
```

To do so we can also reduce the necessary transformation between `BufferAndAvailability` and `BufferResponse`.

wsry replied:
We need ```moreAvailable``` filed to judge whether to add the reader to available queue or not. I replace ```BufferAndAvailability``` with ```ServerOutboundMessage``` and introduce ```build``` method like at client side to build ```BufferResponse``` or ```addBacklog``` message.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -70,6 +70,10 @@
 
+	/** The number of non-event buffers to be announced to the downstream. */
+	@GuardedBy("buffers")
+	private int unannouncedBacklog;
```
The previous `buffersInBacklog` variable should be replaced by this new variable, to avoid maintaining two variables.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultSubpartition.java:
```diff
@@ -158,11 +158,18 @@ public boolean add(BufferConsumer bufferConsumer) throws IOException {
 		private final Buffer buffer;
 		private final boolean isDataAvailable;
 		private final int buffersInBacklog;
+		private final int unannouncedBacklog;
```
`buffersInBacklog` and `unannouncedBacklog` should be retained only one finally.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -94,13 +99,27 @@ public void requestSubpartitionView(
-	public void addCredit(int creditDeltas) {
+	public boolean addCredit(int creditDeltas) {
 		numCreditsAvailable += creditDeltas;
+		return shouldAnnounceBacklog();
```
why we need to trigger announce backlog while adding credit?
I assume since the added `creditDeltas` is always more than zero, then we have the chance to announce the backlog later via sending `BufferResponse`.

wsry replied:
You are right ```shouldAnnounceBacklog``` always return false, I will replace the function call with ```return false```. The reason why we need to change return type of addCredit is that we can't identify whether we are calling ```addCredit``` or ```resumeConsumption``` in ```PartitionRequestQueue```.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputChannel.java:
```diff
@@ -305,19 +311,18 @@ protected CheckpointBarrier parseCheckpointBarrierOrNull(Buffer buffer) throws I
 
-		public BufferAndAvailability(Buffer buffer, boolean moreAvailable, int buffersInBacklog) {
-			this.buffer = checkNotNull(buffer);
+		public BufferAndAvailability(@Nullable Buffer buffer, boolean moreAvailable, int backlog) {
```
We can also avoid introducing `nullable` buffer by this comment https://github.com/apache/flink/pull/11877/files#r422847423

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -94,13 +99,27 @@ public void requestSubpartitionView(
-	public void resumeConsumption() {
+	public boolean resumeConsumption(int availableCredit, int unfulfilledBacklog) {
+		if (initialCredit > 0) {
+			checkState(numCreditsAvailable == availableCredit, "Illegal number of available credit.");
```
I guess this check is not necessary or invalid.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -99,7 +100,7 @@
-	/** The number of required buffers that equals to sender's backlog plus initial credit. */
+	/** The number of buffers to requested that equals to unfulfilled sender's backlog. */
 	@GuardedBy("bufferQueue")
 	private int numRequiredBuffers;
```
nit: might rename to `numRequiredFloatingBuffers` for better reflecting the current semantic.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -127,7 +127,7 @@ private boolean shouldContinueRequest(BufferPool bufferPool) {
 	/**
 	 * Requests exclusive buffers from the provider and returns the number of requested amount.
```
fix the javadoc accordingly.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -122,10 +124,7 @@ public RemoteInputChannel(
 	 * after this input channel is created.
 	 */
 	void assignExclusiveSegments() throws IOException {
-		checkState(initialCredit == 0, "Bug in input channel setup logic: exclusive buffers have " +
```
It is better to supplement the similar check to avoid this action is called multiple times in practice.
Maybe we can check the available exclusive buffers should be 0 in `BufferManager` instead?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java:
```diff
@@ -162,6 +162,10 @@ public void recycle(MemorySegment segment) {
 
 	@Override
 	public List<MemorySegment> requestMemorySegments() throws IOException {
+		if (numberOfSegmentsToRequest <= 0) {
```
`numberOfSegmentsToRequest`  should never be negative because we already check this argument in constructor.
`if (numberOfSegmentsToRequest == 0)` instead?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandler.java:
```diff
@@ -291,6 +290,11 @@ private void decodeMsg(Object msg) throws Throwable {
 					}
 				}
 			}
+		} else if (msgClazz == NettyMessage.AddBacklog.class) {
```
nit: import `NettyMessage.AddBacklog` because it occurs many times in this part.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -129,7 +132,6 @@ private boolean shouldContinueRequest(BufferPool bufferPool) {
 	 */
 	void requestExclusiveBuffers() throws IOException {
 		Collection<MemorySegment> segments = globalPool.requestMemorySegments();
-		checkArgument(!segments.isEmpty(), "The number of exclusive buffers per channel should be larger than 0.");
```
this change should be together with the previous hotfix commit " making `initialCredit` as final"

wsry replied:
Before the last commit, we can't set numExclusive buffers to 0, so the check should be reserved before the last commit.

zhijiangW replied:
Got it. So do you think we add the `if (initialCredit > 0)` before calling this method inside `RemoteInputChannel#assignExclusiveSegments`?  Just for not necessary to `synchronized` below for empty segments.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -139,10 +141,10 @@ void requestExclusiveBuffers() throws IOException {
 	 * requested amount. If the required amount is not fully satisfied, it will register as a listener.
 	 */
-	int requestFloatingBuffers(int numRequired) throws IOException {
+	int requestFloatingBuffers(int backlog) throws IOException {
```
This renaming is not necessary. 

`BufferManager` is abstracted as a general purpose, not coupled with credit-based process, so it is better to not define a argument strongly related to credit-based purpose. `numRequired` seems more general to describe the semantic.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -151,23 +153,63 @@ int requestFloatingBuffers(int numRequired) throws IOException {
+		}
+	}
+
+	public void onCheckpointBarrier(CheckpointBarrier barrier, int initialCredit) {
```
Keep in mind that `BufferManager` is only for buffer request/release related operations, so it should not understand the other specific logics which should be done inside the respective `InputChannel`. Otherwise we would dirty this component and have ambiguous definition what is the role of this component.

In detail, the following should be done inside `RemoteInputChannel`

```
CheckpointOptions options = barrier.getCheckpointOptions();
if (initialCredit == 0 && options.isExactlyOnceMode() && !options.isUnalignedCheckpoint()) 
```
And rename  the method `onCheckpointBarrier` to distinguish with existing `releaseFloatingBuffers()`. From the perspective of outside caller, we should give a clear semantic method naming in order to be reused future.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -151,23 +153,63 @@ int requestFloatingBuffers(int numRequired) throws IOException {
 		return numRequestedBuffers;
 	}
 
+	public NettyMessage.ResumeConsumption resumeAndGetResumptionMessage(
```
As mentioned in https://github.com/apache/flink/pull/11877/commits/708b2ceb97564084900c555d6f38e6ba1174d735#r424167698, this logic should be done inside `RemoteInputChannel`, because the `BufferManager` should not understand the specific logics unless buffer request/release.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -334,6 +335,15 @@ public void resumeConsumption() {
+	/**
+	 * Called by netty thread to request buffers and generate {@link NettyMessage.ResumeConsumption} message.
+	 */
+	public NettyMessage.ResumeConsumption resumeAndGetResumptionMessage() throws IOException {
```
resumeAndGetResumptionMessage() -> getResumeMessage()

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -415,11 +425,8 @@ public Buffer requestBuffer() {
-			notifyCreditAvailable();
-		}
+	public void onSenderBacklog(int backlog) throws IOException {
+		notifyBufferAvailable(bufferManager.requestFloatingBuffers(backlog));
```
Make this change a separate hotfix?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -145,20 +148,20 @@ public void close() throws IOException {
 		NetworkSequenceViewReader reader = allReaders.get(receiverId);
 		if (reader != null) {
-			operation.accept(reader);
+			boolean needAnnounceBacklog = function.apply(reader);
```
Actually the logic for adding credit does not need `needAnnounceBacklog`, so it might bring trouble to understand the logic of addCredit by reusing the common codes here.

One possible solution is to call `enqueueAvailableReader` in reader stack while applying the function, then the `NetworkSequenceViewReader#addCredit` and `NetworkSequenceViewReader#resumeConsumption` can judge the separate conditions before calling `enqueueAvailableReader`.

zhijiangW replied:
To do so we can also avoid adjusting to return unnecessary boolean value for `NetworkSequenceViewReader#addCredit`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -94,10 +96,11 @@ void notifyReaderNonEmpty(final NetworkSequenceViewReader reader) {
-	private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
-		if (reader.isRegisteredAsAvailable() || !reader.isAvailable()) {
+	private void enqueueAvailableReader(final NetworkSequenceViewReader reader, boolean announceBacklog) throws Exception {
+		if (reader.isRegisteredAsAvailable() || (!reader.isAvailable() && !announceBacklog)) {
```
Based on this [comment](https://github.com/apache/flink/pull/11877#discussion_r424353439), we might not need to bring `announceBacklog` argument in this method, to understand all the related processes together.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -45,6 +47,8 @@
 
 	private final PartitionRequestQueue requestQueue;
 
+	private final int initialCredit;
```
Renaming it to a boolean type `withoutExclusiveCredits` seems more direct to understand

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -151,23 +150,43 @@ int requestFloatingBuffers(int numRequired) throws IOException {
+				isWaitingForFloatingBuffers = false;
+			}
+
+			int bufferReleased = bufferQueue.releaseFloatingBuffers();
```
nit: numReleasedBuffers

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -151,23 +149,42 @@ int requestFloatingBuffers(int numRequired) throws IOException {
 		return numRequestedBuffers;
 	}
 
+	public void unregisterBufferListenerAndReleaseFloatingBuffers() {
```
This method should be placed into the below section `Buffer recycle`. 

I think it is better to integrate this method with existing `#releaseFloatingBuffers` to provide a general one, otherwise it might bring confusing to understand the difference among them, especially for the different handle of `numRequiredBuffers`, to make them seem customized logic.

The integration is as below

```
void releaseFloatingBuffers(boolean isTemporaryRelease) {
		synchronized (bufferQueue) {
			if (isWaitingForFloatingBuffers) {
				inputChannel.inputGate.getBufferPool().removeBufferListener(this);
				isWaitingForFloatingBuffers = false;
			}

			int numReleasedBuffers = bufferQueue.releaseFloatingBuffers();
			if (isTemporaryRelease) {
				numRequiredBuffers += numReleasedBuffers;
			} else {
				numRequiredBuffers = 0;
			}
		}
	}
```

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -349,7 +373,7 @@ int unsynchronizedGetFloatingBuffersAvailable() {
 		int addExclusiveBuffer(Buffer buffer, int numRequiredBuffers) {
 			exclusiveBuffers.add(buffer);
-			if (getAvailableBufferSize() > numRequiredBuffers) {
+			if (numRequiredBuffers == 0) {
```
It should not be changed here. If the `numRequiredBuffers` is 0, `getAvailableBufferSize()` must be more than it. If `numRequiredBuffers` is `1` or something else, as long as the `getAvailableBufferSize()` is more than it, we also need to release a floating buffer.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -327,6 +322,14 @@ public void notifyBufferAvailable(int numAvailableBuffers) {
 	}
 
+	@Override
+	public void onCheckpointBarrier(CheckpointBarrier barrier) {
```
make it only private method inside `RemoteInputChannel`, because it is never used outside.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -418,8 +437,8 @@ public Buffer requestBuffer() {
 	 */
-	void onSenderBacklog(int backlog) throws IOException {
-		notifyBufferAvailable(bufferManager.requestFloatingBuffers(backlog + initialCredit));
+	public void onSenderBacklog(int backlog) throws IOException {
```
nit: also adjust the javadoc of this method accordingly.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -313,7 +313,7 @@ int getNumberOfAvailableBuffers() {
 	}
 
-	int unsynchronizedGetExclusiveBuffersUsed() {
+	int unsynchronizedGetExclusiveBuffers() {
```
unsynchronizedGetAvailableExclusiveBuffers?

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -328,4 +325,67 @@ public void operationComplete(ChannelFuture future) throws Exception {
+
+	static class AddBacklogMessage extends ServerOutboundMessage {
+
+		AddBacklogMessage(InputChannelID receiverId, int backlog, boolean moreAvailable) {
```
remove `moreAvailable` argument from the constructor, because it seems strange for `AddBacklogMessage`, then we can give false in below `super` instead.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -94,10 +97,11 @@ void notifyReaderNonEmpty(final NetworkSequenceViewReader reader) {
 	 */
-	private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
-		if (reader.isRegisteredAsAvailable() || !reader.isAvailable()) {
+	void enqueueAvailableReader(final NetworkSequenceViewReader reader, BooleanSupplier condition) throws Exception {
```
Adjust the javadoc accordingly

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -163,8 +172,15 @@ boolean hasBuffersAvailable() {
+		if (backlog > 0) {
+			return new AddBacklogMessage(receiverId, backlog, false);
+		}
+		return null;
```
I guess we should not expect `null` return here. As long as the code path enters `getAddBacklogMessage`, then we should guarantee that the respective backlog should be more than 0. 
Maybe add assert backlog instead?

wsry replied:
It can be null because of redundant data available notification.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -69,6 +73,7 @@
 		this.receiverId = receiverId;
 		this.numCreditsAvailable = initialCredit;
 		this.requestQueue = requestQueue;
+		this.withoutExclusiveCredits = initialCredit > 0;
```
initialCredit  == 0;

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -94,13 +99,22 @@ public void requestSubpartitionView(
 
 	@Override
-	public void resumeConsumption() {
+	public void resumeConsumption(int availableCredit, boolean hasUnfulfilledBacklog) throws Exception {
```
nit: availableCredit -> availableCredits

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -357,11 +359,26 @@ public void resumeConsumption() throws IOException {
     }
 
+    private void onBlockingUpstream() {
+        if (initialCredit == 0) {
```
Why can not we do the same thing for any number of credits not only for `initialCredit == 0`? Or does it slow down the load after resuming?

wsry replied:
We can also release the floating buffers if the initialCredit != 0. I think it does not slow down the load after resuming.

akalash commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedSequenceNumberingViewReader.java:
```diff
@@ -100,8 +106,19 @@ public void addCredit(int creditDeltas) {
+        if (initialCredit == 0) {
+            // reset available credit if no exclusive buffer is available at the
+            // consumer side for all floating buffers must have been released
+            numCreditsAvailable = 0;
```
Do we do it because we know that all floating buffers would be released before the checkpoint when `initialCredit == 0`?

wsry replied:
Yes, you are right.

akalash reviewed (commented):
@wsry thanks for the changes. I don't fully finish the review but I left several questions which should help me to understand the whole idea. In general, my main concern is about the direct manipulation of initialCredit. It seems so fragile for me but I can be wrong. I will take some time to think about how it is safe to do so.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandler.java:
```diff
@@ -340,7 +355,15 @@ private void decodeBufferOrEvent(
+                // recycle the empty buffer directly
+                Buffer buffer = bufferOrEvent.getBuffer();
+                if (buffer != null) {
+                    buffer.recycleBuffer();
```
It is not a bug. The empty buffer will not be sent to the downstream before. However, after this change, the empty buffer will be sent to the downstream task to release the credit already allocated for it. If we do not send the empty buffer to the downstream, the corresponding downstream channel will occupy more buffers than needed.

wsry replied:
Yes, you are right. Previously, ```bufferOrEevnt.getBuffer()``` was always ```null```. And currently, it should be never null. So we do not need to support null buffer now.

wsry replied:
I think I can keep this logic unchanged and release the buffer in decoder

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/PartitionRequestQueue.java:
```diff
@@ -164,6 +168,27 @@ void addCreditOrResumeConsumption(
+     */
+    private void announceBacklog(NetworkSequenceViewReader reader) {
+        int backlog = reader.getRemainingBacklog();
+        if (backlog > 0) {
```
The backlog can be 0. For example, after resumption from checkpoint, if there is no pending data in the subpartition.

pnowojski replied:
Have you tested this @wsry ? After all it seems like if reader is available, it should have `backlog > 0`, shouldn't it?

pnowojski replied:
Secondly, `getRemainingBacklog()` is a very costly operation (additional synchronisation) that I think could have been avoided:

1. Backlog can go up only as a result of `org.apache.flink.runtime.io.network.netty.PartitionRequestQueue#notifyReaderNonEmpty(reader)`. 
2. Backlog can go down, only as a result of polling the data from the `reader`.
3. So instead of using thread safe `reader.getRemainingBacklog()`, we could re-use existing synchronisation in 1. and 2., to maintain `remainingBacklog` in the netty thread (in the `PartitionRequestQueue`.


But it would be even better to completely avoid this check (my comment above).

wsry replied:
You are right, we can do some optimization here.

wsry replied:
I think there is one thing blocking us from reusing the backlog from PartitionRequestQueue#notifyReaderNonEmpty(reader): we have duplicated availability notifications. We may get outdated backlog.

wsry replied:
Maybe we can reuse the return value of is available. Instead of just return a bool flag, we may also return the backlog. I will give it a try.

pnowojski replied:
`NetworkSequenceViewReader#isAvailable()` has the same problem of synchronisation costs.

> I think there is one thing blocking us from reusing the backlog from PartitionRequestQueue#notifyReaderNonEmpty(reader): we have duplicated availability notifications. We may get outdated backlog.

What is the problem?  Keep in mind that spurious notifications shouldn't be that big of a problem. If we sometimes wake up too many times, and we rarely send incorrect backlog information that should be fine, as long as we always over estimate the backlog.

After all even in your version I think there can be a race condition where `PartitionRequestQueue` is notified reader is non empty, you check the `getRemainingBacklog()` and send the `BacklogAnnouncement` message, while before it gets processed by the receiver, some buffer is polled from this reader and backlog goes down to `0`. And as a result, receiver assigns as a credit to a sender that doesn't have any data anymore?

wsry replied:
NetworkSequenceViewReader#isAvailable() has already been called by PartitionRequestQueue#enqueueAvailableReader:
```
private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
if (reader.isRegisteredAsAvailable() || !reader.isAvailable()) {
return;
}
......
}
```
I think we can change it to:
```
    private void enqueueAvailableReader(final NetworkSequenceViewReader reader) throws Exception {
        if (reader.isRegisteredAsAvailable()) {
            return;
        }

        ResultSubpartitionView.AvailabilityWithBacklog availabilityWithBacklog =
                reader.getAvailabilityAndBacklog();
        if (!availabilityWithBacklog.isAvailable()) {
            int backlog = availabilityWithBacklog.getBacklog();
            if (backlog > 0 && reader.needAnnounceBacklog()) {
                announceBacklog(reader, backlog);
            }
            return;
        }
......
}
```
What do you think?

> After all even in your version I think there can be a race condition where PartitionRequestQueue is notified reader is non empty, you check the getRemainingBacklog() and send the BacklogAnnouncement message, while before it gets processed by the receiver, some buffer is polled from this reader and backlog goes down to 0. And as a result, receiver assigns as a credit to a sender that doesn't have any data anymore?

I think there is no such problem. One reason is that ```getRemainingBacklog()``` is in netty thread and we can guarantee the order, the downstream will always process the ```BacklogAnnouncement``` before the buffer. The other reason is that we only announce credit when there is no credit available which means if there is credit available, only buffer with backlog will be sent, if there is no credit available, only ```BacklogAnnouncement``` will be send, buffer will be always wait for credit. It is also not a problem if we announce the same credit twice because we can guarantee the process order.

pnowojski replied:
Yes you are right. `reader.isAvailable()` is already being called. I'm not sure why I've missed that.

`reader.getAvailabilityAndBacklog();` should work fine :)

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionDirectTransferReader.java:
```diff
@@ -91,10 +91,14 @@ public BufferAndBacklog getNextBuffer() throws IOException {
+            nextDataType = Buffer.DataType.DATA_BUFFER;
+        } else if (numDataAndEventBuffers > 0) {
+            nextDataType = Buffer.DataType.EVENT_BUFFER;
+        }
```
An EVENT_BUFFER does not need any credit to be sent. The BoundedBlockingSubpartitionDirectTransferReader Implementation does not distinguish EVENT_BUFFER and DATA_BUFFER for simplicity. However, after this change, we need to distinguish EVENT_BUFFER and DATA_BUFFER because we do not want to allocate any credit at downstream for an EVENT_BUFFER.

pnowojski replied:
But what would be a problem with requesting for a credit for the `EndOfPartitionEvent`? In other words, what's wrong with doing it as it was done previously: always returning `DATA_BUFFER` or `NONE`?

wsry replied:
I think this is just an optimization. Without this change, the backlog announced to the downstream does not include the event and EVENT_BUFFER and DATA_BUFFER are not distinguished, this means the event buffer also need a credit to send. If there is no exclusive buffer, no enough buffer will be allocated for the event, because we only announce the data buffer backlog to the downstream. As a result, some tests will hang for there is no credit for the event buffer. To solve this dead lock, there are two simple ways:
1. Announce both data buffer and event buffer backlog to the downstream, this lead to allocate more buffers than needed, these buffers will be released when the EOF is received at the downstream task.
2. Distinguish EVENT_BUFFER and DATA_BUFFER just like what is doing now.

These choices are both acceptable to me. I chose the second one because EVENT_BUFFER and DATA_BUFFER are distinguished at the downstream task and we can allocate one less buffer.

Both of the choices need to do some change to BoundedBlockingSubpartitionDirectTransferReader. Which one do you prefer? I think both are acceptable for me.

wsry replied:
After rethink about it, the first choice can support more events type in the future and the second choice make the assumption that we only have one event at the end of the data. Maybe the first choice is better?

pnowojski replied:
Let's stay with the option 2 (as you have currently implemented)

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/BoundedBlockingSubpartitionDirectTransferReader.java:
```diff
@@ -110,7 +114,12 @@ private void updateStatistics(Buffer buffer) {
         // We simply assume there are no events except EndOfPartitionEvent for bath jobs,
         // then it has no essential effect to ignore the judgement of next event buffer.
-        return numCreditsAvailable > 0 && numDataAndEventBuffers > 0;
+        return (numCreditsAvailable > 0 || numDataBuffers == 0) && numDataAndEventBuffers > 0;
```
Yes, you are right. At the downstream task, no credit is needed when decoding an event.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -288,9 +288,7 @@ BufferAndBacklog pollBuffer() {
-            }
-
-            while (!buffers.isEmpty()) {
+            } else {
```
After the changes, if the first buffer is finished and empty. The empty buffer will be sent to the downstream to release the allocated credit for it instead of recycled directly.

wsry replied:
If there is only one buffer in the buffer queue, after all data is read, the flushRequested flag will be turned off. Consider the following scenario (correct me if I am wrong):

1. There is an unfinished buffer in the queue and all data is read.
2. Then the buffer is finished but no new data is appended. Note appending data and finish buffer is not an atomic operation.
3. New buffer or event is added to the buffer queue.
4. When polling buffer, an empty buffer is at the head of the queue and we already allocate a credit for it.

akalash replied:
> appending data and finish buffer is not an atomic operation.
It is exactly what I want to emphasize. As I understand, we need the empty buffer only because `appending data and finish buffer is not an atomic operation.` which happens because BufferBuilder has a gap between `commit` and `finish`. So I propose to rewrite BufferBuilder in such a way that if it commits the current state and size == capacity it finishes it immediately(so there is no explicit finish call needed). These changes allow us to avoid all logic with an empty buffer which I think pretty untransparent.

One more time, my proposal:

- Rewriting the `BufferBuilder` in such a way that it will be impossible to have the empty finished buffer(it just requires removing the gap between the last commit and finish).
- Don't send the empty buffer(after changes described above there won't be any empty buffers)

@wsry WDYT?
@pnowojski What is your opinion, is it safe to finish BufferBuilder immediately when it becomes full? I mean not to wait for call explicit `finish` but finish in the `commit` if it is full?

wsry replied:
Does that mean that the BufferBuilder can not finish a partial buffer, for example the last piece of data of the batch execution mode? Besides, there is another case which may lead to the empty buffer: in PipelinedApproximateSubpartition, if a partial record takes more than one buffer in the queue, those partial data will be dropped when recovering.

akalash replied:
Yes, the batch execution mode still requires the manual calling of finish which brokes my idea.  The same about PipelinedApproximateSubpartition. As I understand, it is impossible to know that a partial record will be sent or not before the request the credit.
So I still believe that it would be better to avoid sending the empty buffer if it is possible. But perhaps you right and it will be not so easy to do so in the current implementation, especially because of the scenarios which you described.

pnowojski replied:
Can we measure how many those empty buffers are being sent? Is this an issue of 0.1% more messages or 50%? If it's hard to say or can vary, would it be difficult/would it complicate the code to send the empty buffers only if `#exclusiveBuffers == 0` to make sure that there is no performance regression?

wsry replied:
It is pretty simple to send empty buffer only when  #exclusiveBuffers == 0.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -513,19 +514,20 @@ private void increaseBuffersInBacklog(BufferConsumer buffer) {
+
+            if (flushRequested
+                    || isFinished
+                    || !checkNotNull(buffers.peekLast()).getBufferConsumer().isBuffer()) {
```
In the previous if block in this method, we have already filtered the case when the buffer queue is empty.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -513,19 +514,20 @@ private void increaseBuffersInBacklog(BufferConsumer buffer) {
-        } else {
-            return Math.max(buffersInBacklog - 1, 0);
+        synchronized (buffers) {
+            if (isBlocked || buffers.isEmpty()) {
```
After the channel is blocked, we do not need to allocated any credit at the downstream, because no buffer will be sent. As mentioned in your comment, if the initialCredit is 0, it should return 0. If the initialCredit is not 0, we always have exclusive credits, we still do not need allocate any floating credit. Maybe returning the actual value instead of 0 is a small improvement.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -83,8 +84,13 @@ public BufferManager(
 
     @Nullable
-    Buffer requestBuffer() {
+    Buffer requestBuffer(int initialCredit) {
```
You are right, we can avoid this parameter.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -83,8 +84,13 @@ public BufferManager(
+            // decrease the number of buffers require to avoid the possibility of
+            // allocating more than required buffers after the buffer is taken
+            if (initialCredit == 0) {
+                --numRequiredBuffers;
```
You are right, we should always decrease this value.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -130,11 +136,11 @@ private boolean shouldContinueRequest(BufferPool bufferPool) {
-        checkArgument(
-                !segments.isEmpty(),
-                "The number of exclusive buffers per channel should be larger than 0.");
+        if (numExclusiveBuffers <= 0) {
```
You are right, I will add a checkArgument.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -215,9 +221,15 @@ public void recycle(MemorySegment segment) {
+            buffers = bufferQueue.clearFloatingBuffers();
+        }
+
+        // recycle all buffers out of the synchronization block to avoid dead lock
```
Each RemoteInputChannel has its own BufferManager and bufferQueue, floating buffers recycled by one RemoteInputChannel can be assigned to other RemoteInputChannels directly and the buffer assignment also need to sync the bufferQueue object. Besides, different RemoteInputChannels may recycle buffers simultaneously, which means we may need to lock two bufferQueue objects of different RemoteInputChannels simultaneously in the reverse order. I already encountered this dead lock when running test.

akalash replied:
As I understand from your explanation, it relates to this BufferManager#notifyBufferAvailable. But I still don't get which threads can call the releaseFloatingBuffers simultaneously. It is not so important, but just for curiosity if you have the stacktrace of this deadlock can you share it with me(or just tell me which test)?
Anyway, your changes look correct here.

wsry replied:
It is the new logic added by this PR causes the problem. RemoteInputChannel#onBlockingUpstream will call BufferManager#releaseFloatingBuffers in netty thread and different RemoteInputChannels can have different netty thread.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -265,6 +265,13 @@ public void run() {
+
+        // ignore the empty buffer directly
+        if (buffer.readableBytes() == 0) {
+            buffer.recycleBuffer();
```
Before this change, the buffer is recycled in the pollBuffer method of Subpartition because no empty buffer is sent.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -265,6 +265,13 @@ public void run() {
                 next.getSequenceNumber());
+
+        // ignore the empty buffer directly
+        if (buffer.readableBytes() == 0) {
```
After this change, we can poll an empty buffer from the Subpartition of the upstream task.

wsry commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -357,11 +359,26 @@ public void resumeConsumption() throws IOException {
         checkPartitionRequestQueueInitialized();
 
+        if (initialCredit == 0) {
+            unannouncedCredit.set(0);
```
Doing this in the onBlockingUpstream method can guarantee that after resumeConsumption the unannouncedCredit is in a clean state. Because the unannouncedCredit may get increased after that if there is any new available floating credit.

pnowojski replied:
But how can this floating credit be assigned to this blocked `RemoteInputChannel`? Wouldn't it cause the same deadlock, when floating buffers are assigned to blocked channels and job/task can not make any progress?

It sounds like maybe this should have been handled sooner when trying to increase `unannouncedCredit`?

wsry replied:
This because RemoteInputChannel#notifyBufferAvailable which increases ```unannouncedCredit``` and floating buffer assignment is not an atomic operation. There is a possibility:
1. the floating buffer is assigned;
2. the floating buffer is released because of receiving an event which blocks the channel;
3. the notifyBufferAvailable is called and the unannouncedCredit is increased.

Do you mean we should make RemoteInputChannel#notifyBufferAvailable and floating buffer assignment an atomic operation?

pnowojski replied:
Something sounds wrong here. The race condition that you described above, does it mean that `unannouncedCredit` can be out of sync? That we in reality have released all floating buffers, channel is blocked, but actually `unannouncedCredit > 0`? And it's only fixed after calling `resumeConsumption()`?

wsry replied:
>>> Something sounds wrong here. The race condition that you described above, does it mean that unannouncedCredit can be out of sync? That we in reality have released all floating buffers, channel is blocked, but actually unannouncedCredit > 0? And it's only fixed after calling resumeConsumption()?

Yes, exactly.

>>> And as I understand it, without your change, this problem doesn't exist, as floating buffers are kept assigned to the blocked channel and the unannouncedCredit (or maybe even assigned AddCredit that might have been sent to the upstream node) are consistent with the reality. Also those assigned floating buffers are not used because channel is blocked, but that is not a big issue, because thanks to the exclusive buffers, other channels can make a progress?

That only happens when the exclusive credit is 0. If the exclusive credit is not 0, the allocated floating buffers will not be released and if the exclusive credit is 0, we release the floating buffers allocated to let other channel use them to avoid deadlock, an extreme case is that we only have 1 floating buffer and no exclusive buffer. At downstream, the unannounced credit will be reset, at the upstream, the available credit is also reset to 0 when resume consumption.

pnowojski replied:
Ok, can you at least add a comment explaining why this is set to `0` here?

wsry commented:
Hi @akalash really thanks for the review and comments.

I must admit that this PR is pretty fragmented and is not easy to review. One reason is that the existing implementation relies heavily on the assumption that the exclusive credit is not 0. However, this PR breaks this assumption. Many changes in this PR is to achieve the goal of allocating floating credits at the downstream side according to backlog of the upstream task strictly if there is no exclusive credit. One buffer one credit, no more and no less (event does not take any credit). If this is not guaranteed, credit deadlock may occur.

Thanks again for the review. I will run more tests to guarantees the correctness of this change. (Previously, I already ran some tests including the azure pipeline when setting the exclusive buffer to 0)

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/CreditBasedPartitionRequestClientHandler.java:
```diff
@@ -433,8 +459,8 @@ public void operationComplete(ChannelFuture future) throws Exception {
         public Object buildMessage() {
-            return new AddCredit(
-                    inputChannel.getAndResetUnannouncedCredit(), inputChannel.getInputChannelId());
+            int credits = inputChannel.getAndResetUnannouncedCredit();
+            return credits > 0 ? new AddCredit(credits, inputChannel.getInputChannelId()) : null;
```
Why is this an issue? Is this an independent optimisation or is it relate with the other parts of the PR?

pnowojski replied:
Secondly, return type of the `buildMessage()` is not annotated `@Nullable`, so if we need this code for correctness we need to add this annotation (but it would be better to avoid `null` here)

wsry replied:
It is relevant to this PR, it is because we add the reset logic in ```RemoteInputChannel#resumeConsumption``` method.
```
        if (initialCredit == 0) {
            unannouncedCredit.set(0);
        }
```

wsry replied:
I will add a @Nullable annotation.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NetworkBufferAllocator.java:
```diff
@@ -69,6 +69,9 @@ Buffer allocatePooledNetworkBuffer(InputChannelID receiverId) {
      * @return The un-pooled network buffer.
      */
     Buffer allocateUnPooledNetworkBuffer(int size, Buffer.DataType dataType) {
+        if (size <= 0) {
+            return null;
+        }
```
1. `@Nullable`
2. Do we really need to support `null` here? Can not we return empty `Buffer`?

wsry replied:
This method is used to allocate unpooled buffer for event, the size must be positive. I will replace it with a argument check.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -315,6 +326,17 @@ BufferAndBacklog pollBuffer() {
                 if (buffer.readableBytes() > 0) {
                     break;
                 }
+
+                // if we have an empty finished buffer and the exclusive credit is 0, we just return
+                // the empty buffer so that the downstream task can release the allocated credit for
+                // this empty buffer, this happens in two main scenarios currently:
+                // 1. all data of a buffer builder has been read and after that the buffer builder
+                // is finished
+                // 2. in approximate recovery mode, a partial record takes a whole buffer builder
+                if (buffersPerChannel == 0 && bufferConsumer.isFinished()) {
+                    break;
+                }
+
```
Instead of this check here, can not we add similar check outside of the `while (...)` loop? For example replace:

```
if (buffer == null) {
    return null;
}
```
with something like:
```
if (buffer == null) {
   if (buffersPerChannel == 0) {
       return EMPTY_BUFFER;
    }
    else {
       return null;
    }
}
```
?
That way, we would avoid sending empty buffer if there are still more buffers in the backlog that are already enqueued?

wsry replied:
I am afraid we can not skip the empty buffer is the exclusive credit is 0, because we may already allocate the floating credit for it and we can not just release the floating credit because we may already announce the credit to the upstream.

pnowojski replied:
I was not suggesting that. I was suggesting to:
> we would avoid sending empty buffer if there are still more buffers in the backlog that are already enqueued

wsry replied:
I guess I did not fully understand the comment, could you please explain a bit more? From my understand, even there are multiple buffers in the queue and the first is empty, if we do not send the first empty buffer out, how could we guarantee the downstream tasks are not allocating too many buffers if the exclusive credit is 0? For example, if there is two buffers, the first one is empty and the second one is an event, we may already announce 1 backlog to the downstream task and if we send the event directly, dose that mean we can not release the credit already allocated? Correct me if I am wrong.

pnowojski replied:
If there are two buffers, first is empty the second one can be:
1. non empty buffer
2. empty buffer
3. event

In all cases, instead of sending this first empty buffer with the backlog information, as far as I understand, it should be possible to just send any of the above instead with the updated backlog information. From the downstream node perspective, it should make no difference if we hide this empty buffer. If this is 2nd or 3rd case, we should be able to release the floating buffer regardless of that, shouldn't we?

wsry replied:
Let's maybe focus on the 3rd case first and we assume that the exclusive credit is 0.

1. There are only one data buffer in the queue.
2. Flush triggered.
3. All data of the first buffer is committed but the buffer is still not finished.
4. All data of the buffer is consumed by pollBuffer and the available credit becomes 0.
5. The first buffer is finished, the second event is added and the data available notification is triggered.
6. The upstream announces backlog to the downstream to request a credit.
7.  The upstream receives available credit and start to pollBuffer.
8. Skip the first empty buffer and send the second event.
9. The downstream receive the event but the event does not consume any credit.

Do you mean we should change the current logic and release the floating buffer for event in some cases (including reduce the available credit by 1 at the upstream, currently the available credit is not decreased for event)? If there are multiple empty buffers, should we just skip the first one or should we skip all?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -517,19 +539,20 @@ private void increaseBuffersInBacklog(BufferConsumer buffer) {
-            return buffersInBacklog;
-        } else {
-            return Math.max(buffersInBacklog - 1, 0);
+        synchronized (buffers) {
```
`pollBuffer()` would be acquiring the lock twice, wouldn't it? If you really need to make this method `public` you should keep `private int getBuffersInBacklogUnsafe()` without any synchronisation.

wsry replied:
I will add a private int getBuffersInBacklogUnsafe() method

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -468,6 +485,11 @@ public void onBuffer(Buffer buffer, int sequenceNumber, int backlog) throws IOEx
 
+            if (buffer.getDataType().isBlockingUpstream()) {
+                onBlockingUpstream();
+                checkArgument(backlog == 0, "Illegal number of backlog.");
```
add the `backlog` value to the error message?

pnowojski reviewed (changes requested):
Thanks for this change @wsry . I 've left some comments/questions.

Side note, it seems to me like this last commit is too big and could have been split into smaller commits. For example adding and sending `BacklogAnnouncement` (but not yet using it) could have been extracted. Same applies to clearing floating buffers when blocking consumption (also could probably have been implemented as an independent step).

wsry commented:
@pnowojski Thanks for the review and comments. I will spend some time to split the last commit into smaller ones and update the PR soon.

wsry commented:
Hi @pnowojski , I have split the last commit into smaller ones and updated the PR.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -312,6 +323,16 @@ BufferAndBacklog pollBuffer() {
                     decreaseBuffersInBacklogUnsafe(bufferConsumer.isBuffer());
                 }
 
+                // if we have an empty finished buffer and the exclusive credit is 0, we just return
+                // the empty buffer so that the downstream task can release the allocated credit for
+                // this empty buffer, this happens in two main scenarios currently:
+                // 1. all data of a buffer builder has been read and after that the buffer builder
+                // is finished
+                // 2. in approximate recovery mode, a partial record takes a whole buffer builder
+                if (buffersPerChannel == 0 && bufferConsumer.isFinished()) {
+                    break;
+                }
+
```
I can not seem to respond in the previous thread, so I need to start a new one.

> Let's maybe focus on the 3rd case first and we assume that the exclusive credit is 0.
>
> 1. There are only one data buffer in the queue.
> 2. Flush triggered.
> 3. All data of the first buffer is committed but the buffer is still not finished.
> 4. All data of the buffer is consumed by pollBuffer and the available credit becomes 0.
> 5. The first buffer is finished, the second event is added and the data available notification is triggered.
> 6. The upstream announces backlog to the downstream to request a credit.
> 7. The upstream receives available credit and start to pollBuffer.
> 8. Skip the first empty buffer and send the second event.
> 9. The downstream receive the event but the event does not consume any credit.
>
> Do you mean we should change the current logic and release the floating buffer for event in some cases (including reduce the available credit by 1 at the upstream, currently the available credit is not decreased for event)?

No, but I think we could send this regardless if any is credit available or not as we are doing right now. I think we are also already attaching information about the backlog to such event. One thing to add (unless we are not doing it already) would be to use this backlog information, to maybe release floating buffers if backlog dropped to 0?

> If there are multiple empty buffers, should we just skip the first one or should we skip all?

We could skip all of them, until we reach one of the three options:
1. non empty data buffer
2. event (check above)
3. last empty buffer, without any events after it - here we would indeed need to send that empty buffer

wsry replied:
> One thing to add (unless we are not doing it already) would be to use this backlog information, to maybe release floating buffers if backlog dropped to 0?

Currently, we are not doing that. Actually, I think it is a little complicated to do so. Because we need to keep consistency between the sender side available credit and the receiver side floating buffers. If we just release the floating buffers at the receiver side, if the sender side available credit is not reset, then there is may data sent out without buffers at receiver side to receive them. If we also reset the available credit at the sender side when the backlog is 0, there is a possibility that some AddCredit messages are on the way and we are not resetting this part. Maybe one way is to not sending any data out after sending a buffer with 0 backlog at sender side, then the receivers clear all floating credits and send a reset message to the senders. Then the senders reset all available credits. This process is similar to the channel blocking and resumption. I think this is a little complicated and can incur extra overhead.

What do you think? Or is there any simple way?

pnowojski replied:
Yes, you are right. It wouldn't be that simple. In that case, how complicated would it be to optimise the code to skip the all of the empty buffers until:
1. non empty data buffer
2. event (then send empty buffer first)
3. last empty buffer, without any events after it - here we would indeed need to send that empty buffer

?

wsry replied:
Currently, I can only come up with the following way, which depends on the downstream to reset the available credit of the upstream. This at least needs to add a special network message and propagating this message can incur extra overhead. If you think this is really important, I will spend some time to rethink about it and see if I can find a better way to solve it.

> Maybe one way is to not sending any data out after sending a buffer with 0 backlog at sender side, then the receivers clear all floating credits and send a reset message to the senders. Then the senders reset all available credits. This process is similar to the channel blocking and resumption.

pnowojski replied:
Ok, let's go with the current way for now, as with `buffersPerChannel == 0` check it's not that critical. But it would be nice if you could spend a little time thinking if we can improve this.

wsry replied:
OK, thanks a lot. I will spend some time thinking if we can improve it.

wsry commented:
@pnowojski I have rebased the latest master branch and appended a fixup commit which fixes the comments and enables 0 exclusive credit for unaligned checkpoint itcase.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/ResultPartitionFactory.java:
```diff
@@ -156,15 +155,16 @@ public ResultPartition create(
+                if (type == ResultPartitionType.PIPELINED_APPROXIMATE) {
+                    subpartitions[i] =
+                            new PipelinedApproximateSubpartition(
+                                    i, networkBuffersPerChannel, pipelinedPartition);
```
rename `networkBuffersPerChannel` -> `configuredNetworkBuffersPerChannel` to better reflect that we are actually overriding this value for the output?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java:
```diff
@@ -112,8 +120,11 @@
         super(index, parent);
+
+        checkArgument(buffersPerChannel >= 0, "Buffers per channel must be non-negative.");
+        this.buffersPerChannel = buffersPerChannel;
```
Can we rename this property to something like `receiverExclusiveBuffersPerChannel`?  Because actually this is not the number of `buffersPerChannel` for the sender.

pnowojski commented:
Thanks @wsry for all of the changes and your work with addressing the feedback. I think this looks almost good to me % a couple of renames and we need to update the config option descriptions to describe the actual state that there never can be 0 buffers per channel on the sender side. Currently the variable naming in the code and docs/description are misleading in this regard.

wsry commented:
> Thanks @wsry for all of the changes and your work with addressing the feedback. I think this looks almost good to me % a couple of renames and we need to update the config option descriptions to describe the actual state that there never can be 0 buffers per channel on the sender side. Currently the variable naming in the code and docs/description are misleading in this regard.

@pnowojski Thanks very much for the review and feedback. I will update the PR soon accordingly.

wsry commented:
Hi @pnowojski , I have updated the PR with a fixup commit.

pnowojski commented on docs/layouts/shortcodes/generated/all_taskmanager_network_section.html:
```diff
@@ -30,7 +30,7 @@
             <td style="word-wrap: break-word;">2</td>
             <td>Integer</td>
-            <td>Number of exclusive network buffers to use for each outgoing/incoming channel (subpartition/inputchannel) in the credit-based flow control model. It should be configured at least 2 for good performance. 1 buffer is for receiving in-flight data in the subpartition and 1 buffer is for parallel serialization.</td>
+            <td>Number of exclusive network buffers to use for each outgoing/incoming channel (subpartition/input channel) in the credit-based flow control model. It should be configured at least 2 for good performance. 1 buffer is for receiving in-flight data in the subpartition and 1 buffer is for parallel serialization. The minimum valid value can be configured is 0 and when 0 is configured, t ...
```
Maybe a minor reword to:

> The minimum valid value that can be configured is 0. When 0 buffers-per-channel is configured, the exclusive network buffers used per downstream incoming channel will be 0, but for each upstream outgoing channel, max(1, configured value) will be used. In other words we ensure that, for performance reasons, there is at least one buffer per outgoing channel regardless of the configuration.

pnowojski reviewed (approved):
Sorry for bothering you, but I hope that's the last one nitty comment 😳

wsry commented:
Hi @pnowojski , I have updated the PR.

pnowojski commented:
Thanks @wsry , merging.
